scx_layered: cleanup node spread

likewhatevs opened the pull request:
make node spread determine cpu order in a less-fragile manner (i.e. use topo instead of shifting by number of cpus)

## 9950x
### layered node spread order
![Screenshot from 2025-03-17 12-54-25](https://github.com/user-attachments/assets/6c8cc026-d0ae-43be-9445-5a5516a55f93)


### likwid topo
![Screenshot from 2025-03-17 12-54-38](https://github.com/user-attachments/assets/e29d5360-df14-44fd-a6bf-61ff3dea8e22)

## dual socket 7282
### layered node spread order
![Screenshot from 2025-03-17 12-56-56](https://github.com/user-attachments/assets/29e1e1f4-736e-4cc5-b16b-f82c30788f90)

### likwid topo
![Screenshot from 2025-03-17 12-57-08](https://github.com/user-attachments/assets/2f003808-022c-4c56-8b47-ba9c6daead67)

likewhatevs commented:
not sure why this isn't doing what I'd expect (nor if it's right or if what I expect is right).

htejun commented on scheds/rust/scx_layered/src/layer_core_growth.rs:
```diff
@@ -243,28 +243,23 @@ impl<'a> LayerCoreOrderGenerator<'a> {
+        // XXX this probably does not work with nosmt.
+        // but layered doesn't work with that at all atm.
+
+        for (_, node) in self.topo.nodes.clone() {
```
Here and below, are `.clone()`s necessary? Can't you just `.iter()`?

likewhatevs commented:
ahh ok -- cache topo in likwid makes it clear how this should work:
![image](https://github.com/user-attachments/assets/3be955b5-f686-4b20-a235-52f2f20a7203)

htejun commented on scheds/rust/scx_layered/src/layer_core_growth.rs:
```diff
@@ -243,28 +243,23 @@ impl<'a> LayerCoreOrderGenerator<'a> {
+            for sub_vec in cpus_vec.clone() {
                 if i < sub_vec.len() {
                     cores.push(sub_vec[i]);
                 }
```
Hmm... this looks complicated and still fragile (what if not all nodes have the same number of CPUs?). Wouldn't something like the following work?
```
    let mut cores: Vec<usize> = vec![];
    let mut iters:Vec<_> = self.topo.nodes.iter().map(|x| x.iter()).collect();
    loop {
        let mut had_some = false;
        for it in &mut iters {
            if let Some(v) = it.next() {
                cores.push(*v);
                had_some = true;
            }
        }
        if !had_some {
            break;
        }
    }
```

htejun replied:
Oh, the above assumes that an iterator keeps returning `None` after returning `None` once which isn't guarnateed. Maybe just `.remove()` the iterator from the vec and replace `loop - if { break; }` with `while (iters.len())`?

likewhatevs replied:
I did this in a different but (I think) not fragile way.

htejun commented:
> ahh ok -- cache topo in likwid makes it clear how this should work: ![image](https://private-user-images.githubusercontent.com/12107998/423595217-3be955b5-f686-4b20-a235-52f2f20a7203.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyMzQwMTAsIm5iZiI6MTc0MjIzMzcxMCwicGF0aCI6Ii8xMjEwNzk5OC80MjM1OTUyMTctM2JlOTU1YjUtZjY4Ni00YjIwLWEyMzUtNTJmMmYyMGE3MjAzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE3VDE3NDgzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEzYmViYTA2ZWY1ZjE1YTZlNmExYzg5MDQ1ZTE0ODJkMjIxOTA3NTQzZmEzYTI2OTIzMzBmMjkzMGRmYWFhMjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ErA_SQU15k-YSjrtroNMEMqYAkjenSLrwTToIOIwQg0)

Right, the ordering should iterate llc by llc in each node to ensure that llc's are given out together.

likewhatevs commented:
I *think* this is right wrt/ topology now, but doesn't make a sense wrt/ ordering once idk topological distances are the same.

Should I hack sorting once distances are the same in here *or* open another PR updating topo to have it do that (i.e. have it sort cores by numerical ID once distances are accounted for/all else is equal?

### 7282 dual socket
![Screenshot from 2025-03-17 13-55-12](https://github.com/user-attachments/assets/2db3490a-9756-4e8c-92b5-1d929aec48a4)
### 9950
![Screenshot from 2025-03-17 13-54-13](https://github.com/user-attachments/assets/79af7eb3-39ef-48fb-be5d-911dcc601dc9)

htejun commented:
Hmm... topo should already have those sorted as they're all in `BTreeMap`s which are iterated in key order.

likewhatevs commented:
> Hmm... topo should already have those sorted as they're all in `BTreeMap`s which are iterated in key order.

Yeah I'm messing something up I think, tried sorting them and that did nothing.

likewhatevs commented:
Ok, I think I follow what this needs to do/it does it now:

9950x:
```
19:57:42 [DEBUG] layer: hhvmworkers algo: NodeSpread core order: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
```

7282 dual socket:
```
19:57:23 [DEBUG] layer: hhvmworkers algo: NodeSpread core order: [0, 16, 1, 17, 2, 18, 3, 19, 4, 20, 5, 21, 6, 22, 7, 23, 8, 24, 9, 25, 10, 26, 11, 27, 12, 28, 13, 29, 14, 30, 15, 31]
```

somewhere along the line I mixed up prox map/likwid output's and growth algo. growth algo runs on cores (not cpus like prox map/likwid output), so for this to be right it needs to not account for CPUs at all.

likewhatevs commented:
Hmm ok so this is really starting to really confuse me.

I added debug logs because I kept getting kinda non-sensical core ordering when like, trying to do things the right way.

These logs are printing what I'm getting back from topo:
```
20:50:03 [DEBUG] LLC ID 0 has cores [0, 1, 8, 9, 10, 11, 12, 13]
20:50:03 [DEBUG] LLC ID 1 has cores [2, 3, 4, 5, 6, 7, 14, 15]
20:50:03 [DEBUG] Node ID 0 has cores [0, 1, 8, 9, 10, 11, 12, 13, 2, 3, 4, 5, 6, 7, 14, 15]
```

This directly opposes what likwid says topology is.
![image](https://github.com/user-attachments/assets/2a201561-2713-4c0a-84aa-9c640e20717c)

likewhatevs commented:
Added debug logs to node spread which dump a key of node/llc/core/cpu mappings. I added this because I didn't understand how this core ordering works, and I'm reasonably confident most folks I've seen use layered misunderstand how this works also. 

Wrt/ this misunderstanding, everyone presumes that Linear/Reverse etc. in configs means "Linear/Reverse wrt/ CPUs (maybe w/ some weirdness around SMT)", not Linear/Reverse wrt/ numerical values of Core IDs, which AFAICT doesn't have any meaning outside of like layered/layer growth algo (i.e. vs the meaning that numerical CPU IDs have, for example)?

Per those logs, this does what it is intended to now. Also, make node spread use the rotation logic.

### core order from 9950x
![Screenshot from 2025-03-17 21-18-06](https://github.com/user-attachments/assets/a1c57f01-ec37-42a6-8f32-7b5d262e16d7)

### core order from dual socket
![Screenshot from 2025-03-17 21-21-11](https://github.com/user-attachments/assets/069504d0-c03f-47f7-bb51-8bab62ae3c96)

etsal commented on scheds/rust/scx_layered/src/layer_core_growth.rs:
```diff
@@ -243,33 +243,44 @@ impl<'a> LayerCoreOrderGenerator<'a> {
+                        .collect::<Vec<usize>>()
+                })
+                .collect();
+            max_node_cpus = std::cmp::max(flat_node_vec.len(), max_node_cpus);
```
Is it ever possible we find more CPUs than the node maximum? Why?

etsal reviewed (approved):
LGTM

hodgesds commented:
I think this works now here's a [gist](https://gist.github.com/hodgesds/bee44744206370021620ea57d63a89ab) with a bunch of different growth algos.
